### PR TITLE
feat: GAP-1 — persist training history with metrics

### PR DESCRIPTION
**Pull request by Sickwoman**
## Summary
Fixes GAP-1 from the backend gaps analysis. Training history (loss, accuracy, 
val_loss, val_accuracy) was computed by Keras but completely thrown away after 
every training run. Users had no way to track model improvement over time.

## Problem
```python
# model_run.py — history was computed but never saved
history = model.fit(x_training, y_training, epochs=...) 
# history.history contains loss[], accuracy[], val_loss[], val_accuracy[]
# but we NEVER saved it — it was just garbage collected
```

### New Files
- `app/models/training_run.py` — new `ModelTrainingRun` SQLModel table
- `app/routers/training_run.py` — 3 new REST endpoints

### Modified Files
- `app/services/model_run.py` — captures `model.fit()` history and persists it
- `app/models/ml.py` — adds `training_runs` relationship to `ModelBasic`
- `app/main.py` — registers the new training_run router
- `migrations/env.py` — imports `ModelTrainingRun` for Alembic autogenerate
- `migrations/versions/9e13d23b8149` — creates `model_training_run` table

## New Endpoints
| Method | Endpoint | Description |
|--------|----------|-------------|
| `GET` | `/api/v1/model/{model_id}/training-runs` | List all runs for a model, newest first |
| `GET` | `/api/v1/model/{model_id}/training-run/{run_id}/metrics` | Full epoch-by-epoch curves for a run |
| `POST` | `/api/v1/model/{model_id}/training-run/{run_id}/set-as-best` | Mark a run as the best |

## What Gets Stored Per Run
- **Timing** — started_at, completed_at, duration_seconds
- **Config snapshot** — epochs, batch_size, training_split, optimizer, loss, metric
- **Final results** — final_train_loss, final_train_metric, final_val_loss, final_val_metric
- **Full curves** — epoch_losses[], epoch_metrics[], epoch_val_losses[], epoch_val_metrics[]
- **Status** — in_progress | success | failed | best

## No Breaking Changes
All existing endpoints and behaviour are completely unchanged.

### CODE REVIEW

**Comment by ivantha** [comment]
Closing this in line with the TensorMap `gsoc` label convention. Issue #120 reserves the full scope of the GSoC 2026 student's term, and this PR overlaps with **Phase 3 (Training visualization and callbacks)**. Persisting epoch-by-epoch training history (the `ModelTrainingRun` table and the curve storage in this PR) is the foundation for the structured Socket.IO metrics and real-time loss/accuracy charts that Phase 3 introduces.

Coding for the selected contributor (Kartikey Gupta) begins on 2026-05-25, so this surface is fenced off until then. Thanks for the contribution; if you would like to contribute before or around GSoC, please pick up a `pre-gsoc` labeled issue instead.
